oat-sa/extension-tao-testqti
# Feature/TR-2846/Manage assets TTL for the caching proxy

Related to: https://oat-sa.atlassian.net/browse/TR-2846

Requires: 
 - [x] https://github.com/oat-sa/tao-test-runner-qti-fe/pull/454

**Short description**
Implement a TTL management for the cached item store.

**Context**
Assets may have a TTL set at the creation of the URL. This value is defined in the file `config/tao/websource_NNNNNNNN.conf.php`, where `NNNNNNNN` is a random identifier. However, this value is not forwarded to the client in any way.

When the user stays for too long in front of a test, and then reach out an item having assets not yet in the cache, the browser will issue a `404` error. This is especially visible in the context of pre-caching where the assets URL is part of the cached data.

**Description**
Add a config option to define a typical TTL for the item store in the context of the pre-caching feature. This value is set in the config file [`config/taoQtiTest/config/testRunner.conf.php`](https://github.com/oat-sa/extension-tao-testqti/blob/dca05a44c27e989b6d9b0291de023bdfa562a621/config/default/testRunner.conf.php#L660), and the default is `15 minutes`.

This value is forwarded to the client, then the pre-caching proxy applies it and prune the expired items from the cache.

**How to test**
- Have a test with several items, and at least one with a video player (the ticket contain samples that can be imported).
- Have the pre-caching feature activated (look at the doc in [confluence](https://oat-sa.atlassian.net/wiki/spaces/DLV/pages/1681359008/pre-caching)).
- Take the test, then before moving to the item with the player, wait a few minutes (usually 20mn with the default settings)
- Move to the item with the video, and play the video. Without the fix, the browser will issue a 404 and the player will fallback to reload mode. With the fix, the video should be played back.

**Note:** To reduce the amount of time to wait to see the error or to validate it, you can tweak the TTL in the config file `config/tao/websource_NNNNNNNN.conf.php`. And set the same value in the file `config/taoQtiTest/config/testRunner.conf.php` (when the fix is applied). This value is expressed in seconds.

**Note 2:** For the time being, another issue may rise, if the user stays in front of the item with the video, and tries to play it back after a long wait: the browser may issue a 404 if the video content is not yet in the cache.
